### DISCUSSION TITLE
Target SSDs/NASs from kubernetes storage classes

### DISCUSSION THREAD

**Discussion by LarsKumbier** [rook/rook#12947] [Q&A] [answered]
I'm experimenting with rook-ceph on kubernetes for storage again. I have 3 nodes, which each have 
- a local ssd-partition mounted at `/dev/sda2`, and 
- a NAS-volume at `/dev/sdb`.

In my rook-config, I use all devices, which automatically finds and uses both as OSDs. So far so good.

What I'd like to do is treat them as two separate storage classes in kubernetes - the ssd-pool is fast but small, while the NAS-pool is slow but huge. That way, I could attach IO-intensive pods like databases onto the ssd-storageclass, while size-heavy pods like fileservers would attach to the NAS-storageclass.

Since rook-ceph currently attaches all OSDs to the same pool, I can not target specific ones via a kubernetes storageclass:

```
bash-4.4$ ceph osd tree
ID  CLASS  WEIGHT   TYPE NAME               STATUS  REWEIGHT  PRI-AFF
-1         0.82942  root default
-7         0.27647      host k8s-storage-1
 1    ssd  0.17879          osd.1               up   1.00000  1.00000¹
 5    ssd  0.09769          osd.5               up   1.00000  1.00000²
-3         0.27647      host k8s-storage-2
 2    ssd  0.17879          osd.2               up   1.00000  1.00000¹
 3    ssd  0.09769          osd.3               up   1.00000  1.00000²
-5         0.27647      host k8s-storage-3
 0    ssd  0.17879          osd.0               up   1.00000  1.00000¹
 4    ssd  0.09769          osd.4               up   1.00000  1.00000²

bash-4.4$ ceph osd status
ID  HOST            USED  AVAIL  WR OPS  WR DATA  RD OPS  RD DATA  STATE
 0  k8s-storage-3   196M   182G      0        0       1       16   exists,up¹
 1  k8s-storage-1   374M   182G      0        0       0        0   exists,up¹
 2  k8s-storage-2   261M   182G      0        0       1       90   exists,up¹
 3  k8s-storage-2   121M  99.8G      0        0       0        0   exists,up²
 4  k8s-storage-3   187M  99.8G      0        0       0        0   exists,up²
 5  k8s-storage-1  85.3M  99.9G      0        0       0        0   exists,up²

¹local SSD storage
²remote NAS storage (it's just a 100gb for testing purposes)
```

I did find a possible solution, where I run two separate rook-ceph instances with a `deviceFilter`, so that one rook-ceph-cluster only uses the SSDs, while the other only uses the NAS-volumes. I dislike this solution due to the disadvantage of having to run two rook-ceph-cluster-instances in one kubernetes-cluster.

I found that I can update the storage classes for devices manually: `ceph osd crush set-device-class osd.XXX` - only thing I'm missing now is how to target them separately from a `kubernetes storageclass`. 

Also, since I'll be using cephFs, the metadata service should reside on the SSD pool due to the higher performance - how do I manage that?

Thank you and Best Regards,
Lars

Ps: I cross-posted this here: https://serverfault.com/q/1144193/92444

**Comment by Madhu-1** [comment, marked as answer]
@LarsKumbier You can create cephcluster to specify deviceClasses for set of OSD'S https://github.com/rook/rook/blob/f70b687a6077877e29ab02b06a881ed2d65c3c78/Documentation/CRDs/Cluster/ceph-cluster-crd.md#osd-configuration-settings and use that deviceclasses for cephfs and rbd pools and create a storageclass to point to the rbd pool and cephfs data pools? this takes out the creation of multiple ceph clusters?

**Reply by LarsKumbier** [reply]
That looks exactly like what I was searching for - thank you! :)
Do you by chance know of any way to overwrite the device class in the configuration (since they are incorrectly detected at the moment), so that I don't have to automate correcting them manually?

**Reply by Madhu-1** [reply]
sorry i dont think its possible. As you didn't specify any deviceClass in cephcluster CR by default they are detected as `ssd`. tagging @travisn to confirm on that.

**Reply by LarsKumbier** [reply]
Just to be sure: I'm currently automating the cluster creation from ansible anyway and it's not productive yet, so I can recreate it with different setting files anytime.

**Reply by Madhu-1** [reply]
Yes thats sounds good, Make sure you can define your own deviceclasses for the set of OSD's you want in cephcluster CR and use custom deviceclass name in the BlockPool and Filesystem CR.

**Reply by LarsKumbier** [reply]
I've just found it in here: https://github.com/rook/rook/blob/f70b687a6077877e29ab02b06a881ed2d65c3c78/Documentation/CRDs/Cluster/ceph-cluster-crd.md#storage-class-device-sets
Thanks a lot! :) Would you like to give a canonical answer in the ServerFault-CrossPost and claim the bounty?

**Reply by Madhu-1** [reply]
@LarsKumbier Thank you.

**Reply by LarsKumbier** [reply]
you're welcome, thanks for helping! :)
